File: codility/31.Peaks.java
public class Peaks {
    
}
class Solution {
    //O(N * log(log(N)))
    public int solution(int[] A) {
        // write your code in Java SE 8
        if(A.length< 3){
            return 0;
        }
        int[] isPeek = new int[A.length];
        for(int i = 1;i<A.length -1;i++){
            if(A[i]> Math.max(A[i-1],A[i+1])){
                isPeek[i] =isPeek[i-1] +1;
            } else {
                isPeek[i] =isPeek[i-1];
            }
        }
        isPeek[A.length-1] = isPeek[A.length-2];
        //System.out.println(Arrays.toString(isPeek));
        
        
        // int pace = 1;
        for(int pace = 1;pace <= A.length;pace ++){
            if(A.length % pace != 0){
                continue;
            }
            int j;
            // 判断第一段
            if(isPeek[pace-1]==0){
                continue; 
            }

            // 判断剩余的
            for(j =2 *  pace;j<=A.length;j+=pace){
                
                // 段首 与 前一段 段首比较；只能两段比较
                if(isPeek[j-1] == isPeek[j-pace-1]) {
                    break;
                }
            }
            // 完整匹配，成功
            if(j > A.length ){
                return A.length/pace;
            }
            
        }
        return 0;
    }
}

File: codility/31.Peaks.py
# A non-empty array A consisting of N integers is given.

# A peak is an array element which is larger than its neighbors. More precisely, it is an index P such that 0 < P < N − 1,  A[P − 1] < A[P] and A[P] > A[P + 1].

# For example, the following array A:

#     A[0] = 1
#     A[1] = 2
#     A[2] = 3
#     A[3] = 4
#     A[4] = 3
#     A[5] = 4
#     A[6] = 1
#     A[7] = 2
#     A[8] = 3
#     A[9] = 4
#     A[10] = 6
#     A[11] = 2
# has exactly three peaks: 3, 5, 10.

# We want to divide this array into blocks containing the same number of elements. More precisely, we want to choose a number K that will yield the following blocks:

# A[0], A[1], ..., A[K − 1],
# A[K], A[K + 1], ..., A[2K − 1],
# ...
# A[N − K], A[N − K + 1], ..., A[N − 1].
# What's more, every block should contain at least one peak. Notice that extreme elements of the blocks (for example A[K − 1] or A[K]) can also be peaks, but only if they have both neighbors (including one in an adjacent blocks).

# The goal is to find the maximum number of blocks into which the array A can be divided.

# Array A can be divided into blocks as follows:

# one block (1, 2, 3, 4, 3, 4, 1, 2, 3, 4, 6, 2). This block contains three peaks.
# two blocks (1, 2, 3, 4, 3, 4) and (1, 2, 3, 4, 6, 2). Every block has a peak.
# three blocks (1, 2, 3, 4), (3, 4, 1, 2), (3, 4, 6, 2). Every block has a peak. Notice in particular that the first block (1, 2, 3, 4) has a peak at A[3], because A[2] < A[3] > A[4], even though A[4] is in the adjacent block.
# However, array A cannot be divided into four blocks, (1, 2, 3), (4, 3, 4), (1, 2, 3) and (4, 6, 2), because the (1, 2, 3) blocks do not contain a peak. Notice in particular that the (4, 3, 4) block contains two peaks: A[3] and A[5].

# The maximum number of blocks that array A can be divided into is three.

# Write a function:

# class Solution { public int solution(int[] A); }

# that, given a non-empty array A consisting of N integers, returns the maximum number of blocks into which A can be divided.

# If A cannot be divided into some number of blocks, the function should return 0.

# For example, given:

#     A[0] = 1
#     A[1] = 2
#     A[2] = 3
#     A[3] = 4
#     A[4] = 3
#     A[5] = 4
#     A[6] = 1
#     A[7] = 2
#     A[8] = 3
#     A[9] = 4
#     A[10] = 6
#     A[11] = 2
# the function should return 3, as explained above.

# Write an efficient algorithm for the following assumptions:

# N is an integer within the range [1..100,000];
# each element of array A is an integer within the range [0..1,000,000,000].

# you can write to stdout for debugging purposes, e.g.
# print("this is a debug message")

def solution(A):
    # Implement your solution here
    # Find all the peaks
    peaks = []
    l = len(A)
    for i in range(1, l - 1):
        if A[i] > A[i-1] and A[i] > A[i+1]:
            peaks.append(i)
    # print(peaks) 

    if len(peaks) == 0:
        return 0
    # Start from len(peaks) to 1 
    k = len(peaks) # The number of blocks

    def isValid(peaks, k, l):
        if l % k != 0:
            return False 
        sub_l = l // k  # The length of each block
        pos = 0
        for i in peaks:
            if i <= pos + sub_l - 1 and i >= pos:
                pos += sub_l 
        # print(pos)
        if pos == l :
            return True 
        else:
            return False

    while k > 0:
        # print(k, isValid(peaks, k, l))
        if isValid(peaks, k, l):
            return k  
        else:
            k -= 1
    return 0